File: LeetCode/Medium/0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        return solve(1,n);
    }
private:
    vector<TreeNode*> solve(int start, int end){
        vector<TreeNode*> trees;
        if(start > end){
            trees.push_back(NULL);
            return trees;
        }
        for(int i = start; i<=end; i++){
            vector<TreeNode*> left = solve(start, i-1);
            vector<TreeNode*> right = solve(i+1, end);
            
            for(auto l:left){
                for(auto r:right){
                    TreeNode* cur = new TreeNode(i, l, r);
                    trees.push_back(cur);
                }
            }
        }
        return trees;
    }
};

File: LeetCode/Medium/0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        ans = []
        def solve(start, end):
            if start > end:
                return [None]
            trees = []
            for i in range(start, end+1):
                left = solve(start,i-1)
                right = solve(i+1, end)

                for l in left:
                    for r in right:
                        cur = TreeNode(i)
                        cur.left = l
                        cur.right = r
                        trees.append(cur)
            return trees
        return solve(1,n)